module/VuFind/src/VuFind/Ratings/RatingsService.php

```diff
@@ -0,0 +1,131 @@
+ */
+
+namespace VuFind\Ratings;
+
+use VuFind\Db\Service\RatingsServiceInterface;
+use VuFind\Record\ResourcePopulator;
+use VuFind\RecordDriver\AbstractBase as RecordDriver;
+
+/**
+ * Ratings service
+ *
+ * @category VuFind
+ * @package  Ratings
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org Main Page
+ */
+class RatingsService
+{
+    /**
+     * Cache for rating data
+     *
+     * @var array
+     */
+    protected $ratingCache = [];
+
+    /**
+     * Constructor
+     *
+     * @param RatingsServiceInterface $dbService         Ratings database service
+     * @param ResourcePopulator       $resourcePopulator Resource populator
+     */
+    public function __construct(
+        protected RatingsServiceInterface $dbService,
+        protected ResourcePopulator $resourcePopulator
+    ) {
+    }
+
+    /**
+     * Get rating information for the provided record.
+     *
+     * Returns an array with the following keys:
+     *
+     * rating - average rating (0-100)
+     * count  - count of ratings
+     *
+     * @param RecordDriver $driver Record to look up
+     * @param ?int         $userId User ID, or null for all users
+     *
+     * @return array
+     */
+    public function getRatingData(RecordDriver $driver, ?int $userId = null)
+    {
+        // Cache data since comments list may ask for same information repeatedly:
+        $recordId = $driver->getUniqueId();
+        $source = $driver->getSourceIdentifier();
+        $cacheKey = $recordId . '-' . $source . '-' . ($userId ?? '');
+        if (!isset($this->ratingCache[$cacheKey])) {
+            $this->ratingCache[$cacheKey] = $this->dbService->getForResource($recordId, $source, $userId);
+        }
+        return $this->ratingCache[$cacheKey];
+    }
+
+    /**
+     * Get rating breakdown for the provided record.
+     *
+     * Returns an array with the following keys:
+     *
+     * rating - average rating (0-100)
+     * count  - count of ratings
+     * groups - grouped counts
+     *
+     * @param RecordDriver $driver Record to look up
+     * @param array        $groups Group definition (key => [min, max])
+     *
+     * @return array
+     */
+    public function getRatingBreakdown(RecordDriver $driver, array $groups)
+    {
+        return $this->dbService->getCountsForResource(
+            $driver->getUniqueId(),
+            $driver->getSourceIdentifier(),
+            $groups
+        );
+    }
+
+    /**
+     * Add or update user's rating for the record.
+     *
+     * @param RecordDriver $driver Record associated with rating
+     * @param int          $userId ID of the user posting the rating
+     * @param ?int         $rating The user-provided rating, or null to clear any existing
+     * rating
+     *
+     * @return void
+     */
+    public function saveRating(RecordDriver $driver, int $userId, ?int $rating): void
```

Comment:
I found it confusing to have addOrUpdateRating methods at multiple layers of the code, so I gave this a different name to help differentiate it.